extend to next note

Ash-86 opened the pull request:
Proposal to add a "extend to next note" action. 

- notes within range selection extend up to next note. 
- last note continues to be extended beyond selection box until next note is found. in this case, selection box extended too. 
- if there isn't any note till end of score, the selected note will be extended to end of score 

bug: Applying on triplets seems to erase the triplet number and bracket above the notes. While execution seems correct, undoing displays strange behaviour. 

https://github.com/user-attachments/assets/19e802f1-e674-410a-aa01-6ced0d10531d

triplets example with undoing going bad:

https://github.com/user-attachments/assets/eb395c81-130b-4d54-9020-3c708463cc28

Jojo-Schmitz commented on src/appshell/view/appmenumodel.cpp:
```diff
@@ -369,7 +370,7 @@ MenuItem* AppMenuModel::makeToolsMenu()
          */
         makeSeparator(),
-        makeMenuItem("copy-lyrics-to-clipboard"),
+        makeMenuItem("copy-lyrics-to-clipboard"),        
```
whitespace damage

Ash-86 replied:
thanks, done.

Jojo-Schmitz commented on src/notation/inotationinteraction.h:
```diff
@@ -247,9 +247,10 @@ class INotationInteraction
     virtual void removeSelectedMeasures() = 0;
     virtual void removeSelectedRange() = 0;
-    virtual void removeEmptyTrailingMeasures() = 0;
+    virtual void removeEmptyTrailingMeasures() = 0;    
```
whitespace damage

Ash-86 replied:
done.

Jojo-Schmitz commented on src/notation/internal/notationactioncontroller.cpp:
```diff
@@ -374,8 +374,9 @@ void NotationActionController::init()
+    registerAction("extend-to-next-note", &Interaction::extendToNextNote);
     registerAction("time-delete", &Interaction::removeSelectedRange);
-    registerAction("del-empty-measures", &Interaction::removeEmptyTrailingMeasures);
+    registerAction("del-empty-measures", &Interaction::removeEmptyTrailingMeasures);    
```
whitespace damage

Ash-86 replied:
done.

Jojo-Schmitz commented on src/notation/internal/notationinteraction.h:
```diff
@@ -252,8 +252,9 @@ class NotationInteraction : public INotationInteraction, public muse::Injectable
+    void extendToNextNote() override;
     void removeSelectedMeasures() override;
-    void removeSelectedRange() override;
+    void removeSelectedRange() override;    
```
whitespace damage

Jojo-Schmitz commented on src/notation/internal/notationuiactions.cpp:
```diff
@@ -443,13 +443,19 @@ const UiActionList NotationUiActions::m_actions = {
              TranslatableString("action", "Delete selected measures"),
              IconCode::Code::DELETE_TANK
-             ),
+             ),   
```
whitespace damage

Jojo-Schmitz commented on src/notation/tests/mocks/notationinteractionmock.h:
```diff
@@ -210,9 +210,10 @@ class NotationInteractionMock : public INotationInteraction
     MOCK_METHOD(void, removeSelectedMeasures, (), (override));
     MOCK_METHOD(void, removeSelectedRange, (), (override));
-    MOCK_METHOD(void, removeEmptyTrailingMeasures, (), (override));
+    MOCK_METHOD(void, removeEmptyTrailingMeasures, (), (override));    
```
whitespace damage

Ash-86 replied:
done.

Ash-86 commented:
@cbjeukendrup any chance i could get some feedback on this?

cbjeukendrup commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3315,6 +3316,48 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+    staff_idx_t endStaff = selection().staffEnd();
+
+    for (EngravingItem* el : selection().elements()) {
+        if (el->isNote() || el->isChord()) {
+            Note* on = toNote(el);
```
This is an error if `!el->isNote()` but `el->isChord()`

cbjeukendrup commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3315,6 +3316,48 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+    staff_idx_t endStaff = selection().staffEnd();
+
+    for (EngravingItem* el : selection().elements()) {
+        if (el->isNote() || el->isChord()) {
```
To reduce nesting, you could invert this `if`:
```c++
    if (!el->isNote()) {
        continue;
     }
```

Ash-86 replied:
Done.

cbjeukendrup commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3315,6 +3316,48 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+                    changeCRlen(ocr, Fraction(num, den));
+                }
+                if (!nextChordRest(nextCR)) {
+                    return;
```
`return`, or `break`? (take into account the case that multiple notes are selected)

Ash-86 replied:
Done.

cbjeukendrup commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3315,6 +3316,48 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+                    addChord(r->tick(), r->ticks(), toChord(ocr), true, r->tuplet());
+                } else {
+                    ocr = prevChordRest(toChordRest(r));
+                    auto num = r->ticks().denominator() * ocr->ticks().numerator() + ocr->ticks().denominator() * r->ticks().numerator();
+                    auto den = ocr->ticks().denominator() * r->ticks().denominator();
+                    changeCRlen(ocr, Fraction(num, den));
```
Given that `Fraction` has a `+` operator, isn't is possible to write
```suggestion
                    changeCRlen(ocr, ocr->ticks() + r->ticks());
```
?

Ash-86 replied:
yes, it works! Done.

cbjeukendrup reviewed (commented):
A couple of comments:

Tantacrul commented:
This is a very interesting little idea. For sure, it would not be a capability found in the main menu... but that's simply a placeholder, I assume.

Potential places it could go:

- It could be a single action tool placed in the 'Customise Toolbar' popover, which means users could potentially add it to their top bar if they like. We could create a little icon for this if @bkunda and others agreed it is nice to have.

- Since it's not a new function, it could also simply live in the context menu (for selected notes) and also as a new shortcut.

FWIW: It should definitely not exist in the properties panel because it's a notation action and not a property.

XiaoMigros commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3315,6 +3316,47 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+        ChordRest* ocr = toChordRest(on->chord());
+        ChordRest* nextCR = nextChordRest(ocr);
+
+        while (nextCR->type() == ElementType::REST) {
```
`        while (nextCR->isRest()) {`

Ash-86 replied:
Done. thanks!

Tantacrul commented:
@Ash-86 - I'm happy to test this and suggest the best UX. Can you fix the issue with the MacOS build and then I can provide any needed assets and perhaps a very small little design spec too? Once we're happy with it, we'll merge.

Cheers!

Jojo-Schmitz commented:
Seems the macOS build failure is outside to what @Ash-86 can fix

cbjeukendrup commented:
Yes, that's the same macOS failure as we're seeing all the time these days. I have a PR to fix it, waiting for review. https://github.com/musescore/MuseScore/pull/27910

Tantacrul commented:
Ah, OK. cool. We'll await the fix on our end.

Jojo-Schmitz commented:
@Tantacrul : the macOS version seems to have been built now

Ash-86 commented:
I fixed a few issues with selecting extended notes in list selection, as well as the last measure issue. I also removed the menu item, so anyone who wants to test this can now use the shortcut Alt+W (though you might have a better one).
The triplet issue is still present, however. Could someone point out how to add a chord to a tuplet in a way that ensures undoing doesn't corrupt the score?

sammik commented:
> Could someone point out how to add a chord to a tuplet in a way that ensures undoing doesn't corrupt the score?

I would look, how it is done by "Tie" command in note input mode.

sammik commented:
One thing, which one is intended, upper one?

![image2](https://github.com/user-attachments/assets/c259175c-a99b-47b0-afa8-4fe101e8a86f)

Ash-86 commented:
> One thing, which one is intended, upper one?

The function uses changeCRlen, so first one.

Ash-86 commented:
Hi @Tantacrul, I fixed the tuplets issue and it should now work with all voices as well. Please test it out and let me know if i missed anything.

Ash-86 commented:
Any chance this can still make it to 4.6? Anything else needed?

Ash-86 commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3318,6 +3319,85 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+    Fraction startTick = selection().tickStart();
+    Fraction endTick = selection().tickEnd();
+    staff_idx_t startStaff = selection().staffStart();
+    staff_idx_t endStaff = selection().staffEnd();
```
Done.

Ash-86 commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3323,6 +3324,72 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+                    crts = cr->ticks();
+                }
+            }
+            endTick = cr->tick() + cr->ticks() >= endTick ? cr->tick() + cr->ticks() : endTick;
```
Done.

Ash-86 commented on src/engraving/dom/cmd.cpp:
```diff
@@ -3318,6 +3319,85 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+        Fraction crts = cr->ticks();
+        crList.push_back(cr);
+
+        while ((nextChordRest(cr) && nextChordRest(cr)->isRest()) || (nextChordRest(cr) && nextChordRest(cr)->tick() > (cr->tick() + cr->ticks()))) {
```
nextchordRest is now called 2 times inside of while loop instead of 5. 
Omitted toRest, used ncr instead.

cbjeukendrup commented on src/engraving/editing/cmd.cpp:
```diff
@@ -3345,6 +3346,92 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+            cr = nextChordRest(cr);
+            crList.push_back(cr);
+            endTick = score()->endTick();
+        }
```
Could you explain this block? 
Specifically, why do you add `cr->measure()->ticks()` to the duration?

Ash-86 replied:
@cbjeukendrup  Anything else needed to move forward? Please let me know if anything is unclear or if you’d like me to make changes.

cbjeukendrup replied:
I would like to understand this slightly better, and I was actually planning to investigate it myself but I never have time for that. 

But now that I look at it again, I realise that this case occurs indeed only in _empty_ measures (well, empty in a non-zero voice). In that case, increasing the note length by the length of a measure makes sense, because in this situation the note should already end on the barline. 
But actually the length should not be increased by the length of the measure that the note is in, but rather by the length of the _next_ measure, right? So maybe we should replace `cr->measure() != score()->lastMeasure()` with `cr->measure()->nextMeasure()` (or something like that), and then use `cr->ticks() + cr->measure()->nextMeasure()->ticks()`.

Finally, let's check that it also works in the case that a non-zero-voice note is followed by an empty measure and then a rest; in that case, not only the empty measure should be filled, but also the rest following it.

cbjeukendrup commented on src/engraving/editing/cmd.cpp:
```diff
@@ -3375,6 +3374,108 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+
+        ChordRest* ncr = nextChordRest(cr);
+
+        while (ncr->isRest() || (ncr->tick() > cr->endTick())) { //second condition checks for empty measure between cr and ncr for voices 2,3,4.
```
Do we need to nullptr-check `ncr` here?

Ash-86 replied:
Maybe I didn't understand how that would work, but ChordRest* ncr = nextChordRest(cr) is called before the while loop, so we can't remove the null check before that. Unless we put eh check in the while loop argument, right?
Also, (cr->endTick() != ncr->tick())  block executes only when voices>0 . If we put it in the while argument, we would still need to check it again inside anyway to know which condition we have (ie, regural rest after, or empty measure). Is there something i am missing?

cbjeukendrup replied:
What I'm hoping for, is that there is a way to have just one `while` loop, without having to have additional checks for empty measures before or after the loop. 

In my suggestion I tried to formulate the overall "success" condition, so we'd just loop until that condition is fulfilled. That does mean that we indeed need a check inside the loop to see why we _don't_ have success yet, which means we're checking some conditions twice. But I'm not worried about that because they are very cheap checks (`ncr` nullcheck, `ncr->isChord()`, and `ncr->tick() == cr->endTick()`).

Does that make sense?

cbjeukendrup commented on src/engraving/editing/cmd.cpp:
```diff
@@ -3375,6 +3374,108 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+            if (cr->endTick() != ncr->tick()) { // check for empty measures and fill with mRest for voices>0
+                is.setTrack(cr->track());
+                is.setSegment(cr->measure()->nextMeasure()->segments().firstCRSegment());
+                score()->enterRest(cr->measure()->nextMeasure()->ticks());
```
Instead of first entering a rest and then immediately replacing that rest, it would be nicer to extend the chord directly. Or would that be complicated?

cbjeukendrup replied:
Would it be an idea to put the block in a static function or a lambda function and call that twice?

cbjeukendrup commented on src/engraving/editing/cmd.cpp:
```diff
@@ -3375,6 +3374,108 @@ void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
+                changeCRlen(cr, newDur);
+                Fraction tickAtDurEnd = cr->tick() + newDur;
+                while (nextChordRest(cr) && tickAtDurEnd > nextChordRest(cr)->tick()) { // if resulting cr has ties
+                    cr = nextChordRest(cr);
```
You could add an auxiliary variable here to avoid calling `nextChordRest(cr)` thrice with the same `cr`

Ash-86 commented:
@zacjansheski @cbjeukendrup Before we merge, I wanted to suggest a feature: extending notes from partially selected chords individually instead of always extending the whole chord. This would align with the behavior of R (and soon T) on list selection. I've already made the changes and can push them next week. What are your thoughts?"
